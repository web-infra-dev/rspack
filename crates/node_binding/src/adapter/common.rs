use std::sync::Arc;

use dashmap::DashMap;
use napi::threadsafe_function::{ErrorStrategy, ThreadsafeFunction};
use once_cell::sync::Lazy;
use tokio::sync::oneshot::Sender;

pub type ThreadsafeRspackCallback = ThreadsafeFunction<String, ErrorStrategy::CalleeHandled>;

pub static REGISTERED_DONE_SENDERS: Lazy<Arc<DashMap<usize, Sender<()>>>> =
  Lazy::new(Default::default);

pub static REGISTERED_PROCESS_ASSETS_SENDERS: Lazy<Arc<DashMap<usize, Sender<()>>>> =
  Lazy::new(Default::default);
