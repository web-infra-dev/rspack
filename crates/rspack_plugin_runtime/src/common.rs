use rspack_core::{RuntimeSourceNode, RUNTIME_PLACEHOLDER_INSTALLED_MODULES};

use crate::ChunkHash;

pub fn generate_common_init_runtime() -> RuntimeSourceNode {
  RuntimeSourceNode {
    content: include_str!("runtime/common/_init_runtime.js").to_string(),
  }
}

pub fn generate_common_module_and_chunk_data() -> RuntimeSourceNode {
  RuntimeSourceNode {
    content: include_str!("runtime/common/_module_and_chunk_data.js")
      .to_string()
      .replace(
        "__INSTALLED_MODULES__",
        RUNTIME_PLACEHOLDER_INSTALLED_MODULES,
      ),
  }
}

pub fn generate_common_check_by_id() -> RuntimeSourceNode {
  RuntimeSourceNode {
    content: include_str!("runtime/common/_check_by_id.js").to_string(),
  }
}

pub fn generate_common_public_path(public_path: &str) -> RuntimeSourceNode {
  RuntimeSourceNode {
    content: include_str!("runtime/common/_public_path.js")
      .to_string()
      .replace("__PUBLIC_PATH_PLACEHOLDER__", public_path),
  }
}

pub fn generate_common_dynamic_data(js: Vec<ChunkHash>, css: Vec<ChunkHash>) -> RuntimeSourceNode {
  RuntimeSourceNode {
    content: include_str!("runtime/common/_dynamic_data.js")
      .to_string()
      .replace(
        "__JS__",
        &format!(
          r#"{{{}}}"#,
          js.iter().fold(String::new(), |prev, cur| {
            prev
              + format!(
                r#""{}": "{}","#,
                cur.name,
                if let Some(hash) = &cur.hash {
                  hash.clone()
                } else {
                  String::new()
                }
              )
              .as_str()
          })
        ),
      )
      .replace(
        "__CSS__",
        &format!(
          r#"{{{}}}"#,
          css.iter().fold(String::new(), |prev, cur| {
            prev
              + format!(
                r#""{}": "{}","#,
                cur.name,
                if let Some(hash) = &cur.hash {
                  hash.clone()
                } else {
                  String::new()
                }
              )
              .as_str()
          })
        ),
      ),
  }
}
