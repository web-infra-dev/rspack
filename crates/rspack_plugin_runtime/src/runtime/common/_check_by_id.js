(function () {
	runtime.checkById = function (obj, prop) {
		return Object.prototype.hasOwnProperty.call(obj, prop);
	};
})();
