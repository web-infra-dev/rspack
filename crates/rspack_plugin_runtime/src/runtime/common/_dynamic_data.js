(function () {
	runtime.installedCssChunks = {};
})();

(function () {
	runtime.chunkHashData = {
		js: __JS__,
		css: __CSS__
	};
})();

(function () {
	runtime.setChunkHashData = function (chunkId, hash, type) {
		return (this.chunkHashData[type][chunkId] = hash);
	};
})();

(function () {
	runtime.__rspack_has_dynamic_chunk__ = function (chunkId, type) {
		return Boolean(
			this.chunkHashData &&
				this.chunkHashData[type] &&
				typeof this.chunkHashData[type][chunkId] !== "undefined"
		);
	};
})();
