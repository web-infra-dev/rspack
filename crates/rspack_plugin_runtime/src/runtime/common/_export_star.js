(function () {
	__webpack_require__.es = function (from, to) {
		Object.keys(from).forEach(function (k) {
			if (k !== "default" && !Object.prototype.hasOwnProperty.call(to, k))
				Object.defineProperty(to, k, {
					enumerable: true,
					get: function () {
						return from[k];
					}
				});
		});
		return from;
	};
})();
