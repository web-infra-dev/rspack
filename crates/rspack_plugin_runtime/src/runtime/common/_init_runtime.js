// runtime instance
var runtime = new Object();
self['__rspack_runtime__'] = runtime;