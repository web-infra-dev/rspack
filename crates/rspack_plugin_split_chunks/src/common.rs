use std::ops::{Deref, DerefMut};
use std::sync::Arc;

use derivative::Derivative;
use rspack_core::{Chunk, ChunkGroupByUkey, Module, SourceType};
use rspack_regex::RspackRegex;
use rustc_hash::{FxHashMap, FxHashSet};

pub type ChunkFilter = Arc<dyn Fn(&Chunk, &ChunkGroupByUkey) -> bool + Send + Sync>;
pub type ModuleTypeFilter = Arc<dyn Fn(&dyn Module) -> bool + Send + Sync>;

pub fn create_default_module_type_filter() -> ModuleTypeFilter {
  Arc::new(|_| true)
}

pub fn create_async_chunk_filter() -> ChunkFilter {
  Arc::new(|chunk, chunk_group_db| !chunk.can_be_initial(chunk_group_db))
}

pub fn create_initial_chunk_filter() -> ChunkFilter {
  Arc::new(|chunk, chunk_group_db| chunk.can_be_initial(chunk_group_db))
}

pub fn create_all_chunk_filter() -> ChunkFilter {
  Arc::new(|_chunk, _chunk_group_db| true)
}

pub fn create_chunk_filter_from_str(chunks: &str) -> ChunkFilter {
  match chunks {
    "initial" => create_initial_chunk_filter(),
    "async" => create_async_chunk_filter(),
    "all" => create_all_chunk_filter(),
    _ => panic!("Invalid chunk type: {chunks}"),
  }
}

pub fn create_regex_chunk_filter_from_str(re: RspackRegex) -> ChunkFilter {
  Arc::new(move |chunk, _| chunk.name.as_ref().map_or(false, |name| re.test(name)))
}

#[derive(Debug, Default, Clone)]
pub struct SplitChunkSizes(pub(crate) FxHashMap<SourceType, f64>);

impl SplitChunkSizes {
  pub fn empty() -> Self {
    Self(Default::default())
  }

  pub fn with_initial_value(default_size_types: &[SourceType], initial_bytes: f64) -> Self {
    Self(
      default_size_types
        .iter()
        .map(|ty| (*ty, initial_bytes))
        .collect(),
    )
  }

  /// Port https://github.com/webpack/webpack/blob/c1a5e4fdeef6c64b4f5624830de7abdecba6301a/lib/optimize/SplitChunksPlugin.js#L283-L290
  pub fn merge(mut self, other: &Self) -> Self {
    other.iter().for_each(|(ty, size)| {
      if !self.contains_key(ty) {
        self.insert(*ty, *size);
      }
    });

    self
  }

  pub fn combine_with(&mut self, other: &Self, combine: &impl Fn(f64, f64) -> f64) {
    let source_types = self
      .keys()
      .chain(other.keys())
      .copied()
      .collect::<FxHashSet<_>>();

    source_types.into_iter().for_each(|ty| {
      let self_size = self.get(&ty).copied();
      let other_size = other.get(&ty).copied();
      match (self_size, other_size) {
        (None, Some(size)) | (Some(size), None) => {
          self.insert(ty, size);
        }
        (Some(self_size), Some(other_size)) => {
          self.insert(ty, combine(self_size, other_size));
        }
        (None, None) => {}
      }
    })
  }

  pub fn bigger_than(&self, other: &Self) -> bool {
    self.iter().any(|(ty, ty_size)| {
      if *ty_size == 0.0 {
        false
      } else {
        let Some(other_size) = other.get(ty).copied() else {
          return false;
        };
        *ty_size > other_size
      }
    })
  }
  pub fn smaller_than(&self, other: &Self) -> bool {
    self.iter().any(|(ty, ty_size)| {
      if *ty_size == 0.0 {
        false
      } else {
        let Some(other_size) = other.get(ty).copied() else {
          return false;
        };
        *ty_size < other_size
      }
    })
  }

  pub fn add_by(&mut self, other: &Self) {
    self.combine_with(other, &|a, b| a + b)
  }
}

impl Deref for SplitChunkSizes {
  type Target = FxHashMap<SourceType, f64>;

  fn deref(&self) -> &Self::Target {
    &self.0
  }
}

impl DerefMut for SplitChunkSizes {
  fn deref_mut(&mut self) -> &mut Self::Target {
    &mut self.0
  }
}

#[derive(Derivative)]
#[derivative(Debug)]
pub struct FallbackCacheGroup {
  #[derivative(Debug = "ignore")]
  pub chunks_filter: ChunkFilter,
  pub min_size: SplitChunkSizes,
  pub max_async_size: SplitChunkSizes,
  pub max_initial_size: SplitChunkSizes,
  pub automatic_name_delimiter: String,
}
