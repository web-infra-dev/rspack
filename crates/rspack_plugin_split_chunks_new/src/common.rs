use std::{
  future::{self, Future},
  ops::{Deref, DerefMut},
  pin::Pin,
  sync::Arc,
};

use derivative::Derivative;
use futures_util::FutureExt;
use rspack_core::{Chunk, ChunkGroupByUkey, Module, SourceType};
use rustc_hash::{FxHashMap, FxHashSet};

pub type ChunkFilter = Arc<dyn Fn(&Chunk, &ChunkGroupByUkey) -> bool + Send + Sync>;

pub fn create_async_chunk_filter() -> ChunkFilter {
  Arc::new(|chunk, chunk_group_db| !chunk.can_be_initial(chunk_group_db))
}

pub fn create_initial_chunk_filter() -> ChunkFilter {
  Arc::new(|chunk, chunk_group_db| chunk.can_be_initial(chunk_group_db))
}

pub fn create_all_chunk_filter() -> ChunkFilter {
  Arc::new(|_chunk, _chunk_group_db| true)
}

pub fn create_chunk_filter_from_str(chunks: &str) -> ChunkFilter {
  match chunks {
    "initial" => create_initial_chunk_filter(),
    "async" => create_async_chunk_filter(),
    "all" => create_all_chunk_filter(),
    _ => panic!("Invalid chunk type: {chunks}"),
  }
}

pub type ModuleFilter = Arc<dyn Fn(&dyn Module) -> bool + Send + Sync>;

fn create_default_module_filter() -> ModuleFilter {
  Arc::new(|_| true)
}

pub fn create_module_filter_from_rspack_regex(re: rspack_regex::RspackRegex) -> ModuleFilter {
  Arc::new(move |module| {
    module
      .name_for_condition()
      .map_or(false, |name| re.test(&name))
  })
}

pub fn create_module_filter(re: Option<String>) -> ModuleFilter {
  re.map(|test| {
    let re =
      rspack_regex::RspackRegex::new(&test).unwrap_or_else(|_| panic!("Invalid regex: {}", &test));
    create_module_filter_from_rspack_regex(re)
  })
  .unwrap_or_else(create_default_module_filter)
}

#[derive(Debug, Default, Clone)]
pub struct SplitChunkSizes(pub(crate) FxHashMap<SourceType, f64>);

impl SplitChunkSizes {
  pub fn empty() -> Self {
    Self(Default::default())
  }

  pub fn with_initial_value(default_size_types: &[SourceType], initial_bytes: f64) -> Self {
    Self(
      default_size_types
        .iter()
        .map(|ty| (*ty, initial_bytes))
        .collect(),
    )
  }

  /// Port https://github.com/webpack/webpack/blob/c1a5e4fdeef6c64b4f5624830de7abdecba6301a/lib/optimize/SplitChunksPlugin.js#L283-L290
  pub fn merge(mut self, other: &Self) -> Self {
    other.iter().for_each(|(ty, size)| {
      if !self.contains_key(ty) {
        self.insert(*ty, *size);
      }
    });

    self
  }

  pub fn combine_with(&mut self, other: &Self, combine: &impl Fn(f64, f64) -> f64) {
    let source_types = self
      .keys()
      .chain(other.keys())
      .copied()
      .collect::<FxHashSet<_>>();

    source_types.into_iter().for_each(|ty| {
      let self_size = self.get(&ty).copied();
      let other_size = other.get(&ty).copied();
      match (self_size, other_size) {
        (None, Some(size)) | (Some(size), None) => {
          self.insert(ty, size);
        }
        (Some(self_size), Some(other_size)) => {
          self.insert(ty, combine(self_size, other_size));
        }
        (None, None) => {}
      }
    })
  }

  pub fn bigger_than(&self, other: &Self) -> bool {
    self.iter().any(|(ty, ty_size)| {
      if *ty_size == 0.0 {
        false
      } else {
        let other_size = other.get(ty).copied().unwrap_or(-f64::INFINITY);
        *ty_size > other_size
      }
    })
  }
  pub fn smaller_than(&self, other: &Self) -> bool {
    self.iter().any(|(ty, ty_size)| {
      if *ty_size == 0.0 {
        false
      } else {
        let other_size = other.get(ty).copied().unwrap_or(f64::INFINITY);
        *ty_size < other_size
      }
    })
  }

  pub fn add_by(&mut self, other: &Self) {
    self.combine_with(other, &|a, b| a + b)
  }
}

impl Deref for SplitChunkSizes {
  type Target = FxHashMap<SourceType, f64>;

  fn deref(&self) -> &Self::Target {
    &self.0
  }
}

impl DerefMut for SplitChunkSizes {
  fn deref_mut(&mut self) -> &mut Self::Target {
    &mut self.0
  }
}

type PinFutureBox<T> = Pin<Box<dyn Future<Output = T> + Send>>;

pub type ChunkNameGetter = Arc<dyn Fn(&dyn Module) -> PinFutureBox<Option<String>> + Send + Sync>;

pub fn create_chunk_name_getter_by_const_name(name: String) -> ChunkNameGetter {
  Arc::new(move |_module| future::ready(Some(name.clone())).boxed())
}

pub fn create_empty_chunk_name_getter() -> ChunkNameGetter {
  Arc::new(move |_module| future::ready(None).boxed())
}

#[derive(Derivative)]
#[derivative(Debug)]
pub struct FallbackCacheGroup {
  #[derivative(Debug = "ignore")]
  pub chunks_filter: ChunkFilter,
  pub min_size: SplitChunkSizes,
  pub max_async_size: SplitChunkSizes,
  pub max_initial_size: SplitChunkSizes,
}
