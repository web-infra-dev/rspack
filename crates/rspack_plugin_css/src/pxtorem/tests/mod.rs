use hashbrown::HashMap;
use hrx_parser::Entry;
use std::fmt::Write;
use swc_core::common::{FileName, FilePathMapping, SourceMap};
use swc_css::{
  ast::Stylesheet,
  codegen::{
    writer::basic::{BasicCssWriter, BasicCssWriterConfig},
    CodeGenerator, CodegenConfig, Emit,
  },
  parser::{parse_file, parser::ParserConfig},
  visit::VisitMutWith,
};

use super::{option::PxToRemOption, px_to_rem::px_to_rem};

#[test]
fn valid() {
  let file = include_str!("./cases/fixtures.hrx");
  for unit in normalize(file)
    .into_iter()
    .filter(|unit| !unit.path.starts_with('-'))
  {
    insta::with_settings!({sort_maps => false, snapshot_path => "cases", prepend_module_to_snapshot => false, snapshot_suffix => ""}, {
      let config = unit.meta_data.get("config");
      let expected = unit.meta_data.get("expected").cloned().unwrap_or_default();
      let snapshot_result = get_snapshot_result(&unit.content, &expected, &transform(&unit.content, config));
      let snapshot_path = unit.path.replace(' ', "_");
      insta::assert_snapshot!(snapshot_path.clone(), snapshot_result, &snapshot_path);
    });
  }
}

fn transform(source: &str, config_file: Option<&String>) -> String {
  let cm = SourceMap::new(FilePathMapping::empty());
  let fm = cm.new_source_file(FileName::Custom("test.css".to_owned()), source.to_owned());
  let mut stylesheet = parse_file::<Stylesheet>(&fm, ParserConfig::default(), &mut vec![]).unwrap();

  let mut output = String::new();
  let wr = BasicCssWriter::new(
    &mut output,
    None, // Some(&mut src_map_buf),
    BasicCssWriterConfig::default(),
  );
  let config: PxToRemOption = config_file
    .map(|file| serde_json::from_str(file).unwrap())
    .unwrap_or_default();
  let mut gen = CodeGenerator::new(wr, CodegenConfig { minify: false });

  stylesheet.visit_mut_with(&mut px_to_rem(config));
  gen.emit(&stylesheet).unwrap();

  output
}

#[derive(Debug)]
struct TestUnit {
  path: String,
  content: String,
  meta_data: HashMap<String, String>,
}

fn normalize(source: &str) -> Vec<TestUnit> {
  let mut res = vec![];
  let archive = hrx_parser::parse(source).unwrap();
  let mut i = 0;
  while i < archive.entries.len() {
    let entry = &archive.entries[i];
    let mut unit = if let Some(unit) = convert_entry_to_unit(entry) {
      unit
    } else {
      continue;
    };
    i += 1;
    while i < archive.entries.len() {
      let another_entry = &archive.entries[i];
      if !another_entry.path().starts_with('.') {
        break;
      }
      if let Some(another_unit) = convert_entry_to_unit(another_entry) {
        unit
          .meta_data
          .insert(another_unit.path[1..].to_string(), another_unit.content);
      }
      i += 1;
    }
    res.push(unit);
  }
  res
}

fn convert_entry_to_unit(entry: &Entry) -> Option<TestUnit> {
  entry.content().map(|content| {
    let path = entry.path();
    TestUnit {
      path,
      content,
      meta_data: HashMap::default(),
    }
  })
}

fn get_snapshot_result(input: &str, expected: &str, actual: &str) -> String {
  let mut result = String::new();
  writeln!(result, "# Input").unwrap();
  writeln!(result, "{}", input).unwrap();
  writeln!(result, "# Expected").unwrap();
  writeln!(result, "{}", expected).unwrap();
  writeln!(result, "# Actual").unwrap();
  writeln!(result, "{}", actual).unwrap();
  result
}
