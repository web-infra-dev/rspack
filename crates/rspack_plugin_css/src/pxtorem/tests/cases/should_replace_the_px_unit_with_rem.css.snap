---
source: crates/rspack_plugin_css/src/pxtorem/tests/mod.rs
assertion_line: 29
expression: should_replace_the_px_unit_with_rem.css
---
# Input
.rule { font-size: 15px }

# Expected
.rule { font-size: 0.9375rem }


# Actual
.rule {
  font-size: 0.9375rem;
}

